kaldb/src/main/java/com/slack/kaldb/clusterManager/SnapshotDeletionService.java

```diff
@@ -0,0 +1,212 @@
+   * Separate services are responsible for the eviction and deletion of expired replicas, and this
+   * service only is expected to remove snapshots that are not being currently served.
+   *
+   * <p>Consideration should be taken for when configs are changed, such as shortening the replica
+   * lifespan. The expectation is that previously created replicas that may have a longer expiration
+   * would see that respected throughout the duration of their lifespan.
+   *
+   * <p>Deletion of expired snapshots should gracefully handle the scenario where immediately prior
+   * to reaching the end of the configured snapshot lifespan, an on-demand request is received. In
+   * this scenario a new replica would have been created, and the object would be currently
+   * downloading for servicing the request. The expectation from the end user would be that the
+   * request would still be served, as when the request was made it was still "in compliance" with
+   * the lifespan configuration.
+   *
+   * <p>When creating on-demand replicas, care should be taken to not create replicas for snapshots
+   * that already exceed their maximum lifespan to prevent a race condition between on-demand
+   * creation and snapshot deletion.
+   */
+  protected int deleteExpiredSnapshotsWithoutReplicas() {
+    Timer.Sample deletionTimer = Timer.start(meterRegistry);
+
+    Set<String> snapshotIdsWithReplicas =
+        replicaMetadataStore
+            .getCached()
+            .stream()
+            .map(replicaMetadata -> replicaMetadata.snapshotId)
+            .filter(snapshotId -> snapshotId != null && !snapshotId.isEmpty())
+            .collect(Collectors.toUnmodifiableSet());
+
+    long expirationCutoff =
+        Instant.now()
+            .minus(
+                managerConfig.getSnapshotDeletionServiceConfig().getSnapshotLifespanMins(),
+                ChronoUnit.MINUTES)
+            .toEpochMilli();
+    AtomicInteger successCounter = new AtomicInteger(0);
+    List<ListenableFuture<?>> deletedSnapshotList =
+        snapshotMetadataStore
+            .getCached()
+            .stream()
+            // only snapshots that only contain data prior to our cutoff, and have no replicas
+            .filter(
+                snapshotMetadata ->
+                    snapshotMetadata.endTimeUtc < expirationCutoff
+                        && !snapshotIdsWithReplicas.contains(snapshotMetadata.name))
+            .map(
+                snapshotMetadata -> {
+                  ListenableFuture<?> future =
+                      Futures.submit(
+                          () -> {
+                            // we first delete the snapshot reference - if for some reason we fail
+                            // to delete the object from the object storage, the orphaned object
+                            // cleanup service will attempt to delete it later
+                            snapshotMetadataStore.deleteSync(snapshotMetadata);
```

Comment:
I think it may be better to delete this logic in 2 separate steps. Delete the metadata and then delete the S3 assets. Deleting assets in parallel may be subject to some rate limits, so it may be better to do it sequentially.

Reply:
As discussed, I went ahead and added the rate limiting to this so that we can reduce spikes on calls to the metadata and s3 services. This logic is fault tolerant of s3 failures, as the orphaned object cleanup will ensure that any failed deletes to s3 are retried until they are brought into compliance.